lib/build.ml

```diff
@@ -1,30 +1,88 @@
 open Current.Syntax
+open Lwt.Infix
+
+module Raw = Current_docker.Raw
+
+module Op = struct
+  type t = Builder.t
+
+  let id = "ci-build"
+
+  module Key = struct
+    type t = {
+      commit : Current_git.Commit.t;            (* The source code to build and test *)
+      repo : Current_github.Repo_id.t;          (* Used to choose a build cache *)
+      base : Raw.Image.t;                       (* The image with the OCaml compiler to use. *)
+      variant : string;                         (* Added as a comment in the Dockerfile *)
+      analysis : Analyse.Analysis.t;
+    }
+
+    let digest_analysis x =
+      let s = Analyse.Analysis.to_yojson x |> Yojson.Safe.to_string in
+      `String (Digest.string s |> Digest.to_hex)
+
+    let to_json { commit; analysis; base; variant; repo } =
+      `Assoc [
+        "commit", `String (Current_git.Commit.id commit);
+        "analysis", digest_analysis analysis;
+        "base", `String (Raw.Image.digest base);
+        "variant", `String variant;
+        "repo", `String (Fmt.to_to_string Current_github.Repo_id.pp repo);
+      ]
+
+    let digest t = Yojson.Safe.to_string (to_json t)
+  end
+
+  module Value = Current.Unit
+
+  let or_raise = function
+    | Ok () -> ()
+    | Error (`Msg m) -> raise (Failure m)
+
+  let build { Builder.docker_context; pool; build_timeout } job { Key.commit; base; analysis; variant; repo } =
+    let dockerfile =
+      let base = Raw.Image.hash base in
+      Dockerfile.string_of_t (
+        if Analyse.Analysis.is_duniverse analysis then
+          Duniverse_build.dockerfile ~base ~repo ~variant
+        else
+          Opam_build.dockerfile ~base ~info:analysis ~variant
+      )
+    in
+    Current.Job.log job "@[<v2>Using Dockerfile:@,%a@]" Fmt.lines dockerfile;
```

Comment:
Would it be possible to log the Dockerfile in a cut-and-paste-friendly way? A minor nit, but the intending makes it difficult :-)

Reply:
Done - how does this look?

https://ci.ocamllabs.io/github/ocurrent/ocaml-ci/commit/40a6e807b83fe183365e66baad636bed39a9e93b/variant/4.10